pkg/vetter/conflictingvirtualservicehost/vet.go

```diff
@@ -48,63 +79,323 @@ type hostAndGateway struct {
 	hostname string
 }
 
+type routeRule struct {
+	ruleType  routeRuleType
+	route     string
+	vsName    string
+	namespace string
+}
+
+type routeTrie struct {
+	subRoutes  map[string]*routeTrie
+	regexs     []routeRule
+	routeRules []routeRule
+}
+
 type VirtualSvcByHostAndGateway map[hostAndGateway][]*v1alpha3.VirtualService
 
+func asString(rrType routeRuleType) string {
+	if rrType == prefix {
+		return "prefix"
+	} else if rrType == exact {
+		return "exact"
+	} else {
+		return "regex"
+	}
+}
+
+func unwrapNote(note conflictingVsNote, hosts []string) *apiv1.Note {
+	return &apiv1.Note{
+		Type:    note.Type,
+		Summary: note.Summary,
+		Msg:     note.Msg,
+		Level:   apiv1.NoteLevel_ERROR,
+		Attr: map[string]string{
+			"vs_names": note.vsNames,
+			"host":     strings.Join(hosts, " "),
+			"routes":   note.routes,
+		}}
+
+}
+
 // CreateVirtualServiceNotes checks for multiple vs defining the same host and
 // generates notes for these cases
 func CreateVirtualServiceNotes(virtualServices []*v1alpha3.VirtualService) ([]*apiv1.Note, error) {
-	vsByHostAndGateway := VirtualSvcByHostAndGateway{}
+	vsByHost := map[string][]*v1alpha3.VirtualService{}
 	for _, vs := range virtualServices {
 		for _, host := range vs.Spec.GetHosts() {
 			h, err := util.ConvertHostnameToFQDN(host, vs.Namespace)
 			if err != nil {
 				fmt.Printf("Unable to convert hostname: %s\n", err.Error())
 				return nil, err
 			}
-
-			// One VS can have multiple hosts and gateways. Make 1 key per
-			// combination.
-			hg := hostAndGateway{hostname: h}
-			if len(vs.Spec.GetGateways()) > 0 {
-				for _, g := range vs.Spec.GetGateways() {
-					hg.gateway = g
-					populateVirtualServiceMap(hg, vs, vsByHostAndGateway)
-				}
+			if _, ok := vsByHost[h]; !ok {
+				vsByHost[h] = []*v1alpha3.VirtualService{vs}
 			} else {
-				hg.gateway = defaultGateway
-				populateVirtualServiceMap(hg, vs, vsByHostAndGateway)
+				vsByHost[h] = append(vsByHost[h], vs)
 			}
 		}
 	}
 
 	// create vet notes
+
+	// We only want to report the unique hosts for a given conflict.
+	// This should be thought of as a hash map from notes to a set of
+	// host names.
+	notesToUniqueHost := map[conflictingVsNote]map[string]struct{}{}
 	notes := []*apiv1.Note{}
-	for key, vsList := range vsByHostAndGateway {
+	for host, vsList := range vsByHost {
 		if len(vsList) > 1 {
-			// there are multiple vs defining a host
-			vsNames := []string{}
+			conflictingRules, err := conflictingVirtualServices(vsList)
+			if err != nil {
+				return notes, err
+			}
 			for _, vs := range vsList {
-				vsName := vs.Name + "." + vs.Namespace
-				vsNames = append(vsNames, vsName)
-
+				gateways := vs.Spec.GetGateways()
```

Comment:
Let's consider adding this as a separate smaller vetter that just checks for missing GWs